src/android_utils.py

```diff
@@ -107,6 +114,89 @@ def get_home_folder():
     return os.path.join(kolibri_home_file.toString(), "KOLIBRI_DATA")
 
 
+def get_endless_key_paths():
+    def _get_directory_path(volume):
+        if SDK_INT < 30:
+            uuid = volume.getUuid()
+            if uuid is None:
+                return None
+            return os.path.join("/storage", uuid)
+        else:
+            directory_file = volume.getDirectory()
+            if directory_file is None:
+                return None
+            return directory_file.toString()
+
+    storageManager = get_activity().getSystemService(Context.STORAGE_SERVICE)
+    volumesList = storageManager.getStorageVolumes()
+    for i in range(volumesList.size()):
+        volume = volumesList.get(i)
+        state = volume.getState()
+        is_removable = volume.isRemovable()
+        directory_path = _get_directory_path(volume)
+        logger.debug(
+            f"Found volume UUID: {volume.getUuid()} state: {state} "
+            f" is removable: {is_removable} mount path: {directory_path}"
+        )
+        if is_removable and state == "mounted" and directory_path is not None:
+            kolibri_data_path = os.path.join(directory_path, "KOLIBRI_DATA")
+            content_path = os.path.join(kolibri_data_path, "content")
+            db_path = os.path.join(
+                kolibri_data_path, "preseeded_kolibri_home", "db.sqlite3"
+            )
+            if os.path.exists(content_path) and os.path.exists(db_path):
+                return {"content_path": content_path, "db_path": db_path}
+    return None
+
+
+def provision_endless_key_database(endless_key_paths):
+    if endless_key_paths is not None:
+        home_folder = get_home_folder()
+        if os.path.exists(os.path.join(home_folder, "db.sqlite3")):
+            logger.debug("EK database already exists, skipping.")
+            return
+        if not os.path.exists(home_folder):
+            os.mkdir(home_folder)
+        if Environment.isExternalStorageManager():
```

Comment:
Looks like you missed a spot :)

```suggestion
        if has_all_files_access():
```

With this change, the app starts correctly for the first time on my Chromebook with Android 9, using an Endless Key device connected via a USB-C to USB-A adapter.

And the content is there. Yay!

Reply:
Ah thanks for spotting! I was trying with the Android 9 emulator but without a EK USB connected to it, so I haven't tested this part. Glad to know it even works with this change.